File: adk/go/weil_go/collections/set.go
package collections

import (
	"encoding/json"
	"fmt"

	"github.com/weilliptic-public/wadk/adk/go/weil_go/errors"
	"github.com/weilliptic-public/wadk/adk/go/weil_go/runtime"
)

type __unit__ struct{}

// WeilSet is a set collection that stores unique values of comparable type.
// It is implemented internally using a WeilMap with unit values.
type WeilSet[V comparable] struct {
	InnerMap WeilMap[V, __unit__] `json:"map"`
}

// NewWeilSet creates a new WeilSet with the given state identifier.
func NewWeilSet[V comparable](stateId WeilId) *WeilSet[V] {
	return &WeilSet[V]{
		InnerMap: WeilMap[V, __unit__]{stateId},
	}
}

// BaseStatePath returns the base path for this set in the state tree.
func (m *WeilSet[V]) BaseStatePath() string {
	return m.InnerMap.BaseStatePath()
}

// StateTreeKey generates a state tree key for the given value suffix.
func (m *WeilSet[V]) StateTreeKey(suffix *V) string {
	serializedSuffix, _ := json.Marshal(suffix)
	return fmt.Sprintf("%s_%s", m.BaseStatePath(), string(serializedSuffix))
}

// Insert adds a value to the set.
// If the value already exists, this operation has no effect.
func (s *WeilSet[V]) Insert(val *V) {
	runtime.WriteCollectionEntry[V]([]byte(s.StateTreeKey(val)), nil)
}

// Contains checks whether the given value exists in the set.
// Returns true if the value is present, false otherwise.
func (s *WeilSet[V]) Contains(val *V) bool {
	_, err := runtime.ReadCollection[V]([]byte(s.StateTreeKey(val)))
	if err != nil {
		switch err.(type) {
		case *errors.KeyNotFoundInCollectionError:
			return false
		default:
			panic(fmt.Sprintf("error other than `KeyNotFoundInCollection` received while `ReadCollection`: %v", err.Error()))
		}
	}

	return true
}

// Remove removes the given value from the set.
// Returns true if the value was removed, false if it was not in the set.
func (s *WeilSet[V]) Remove(value *V) (bool, error) {
	_, err := runtime.DeleteCollectionEntry[V]([]byte(s.StateTreeKey(value)))
	if err != nil {
		switch err.(type) {
		case *errors.KeyNotFoundInCollectionError:
			return false, nil
		default:
			panic(fmt.Sprintf("error other than `KeyNotFoundInCollection` received while `ReadCollection`: %v", err.Error()))
		}
	}

	return true, nil
}

File: adk/asc/weil-sdk/assembly/collections/set.ts
import { WeilId } from './weil-id'
import { Memory } from '../runtime'
import { JSON } from 'json-as'

/** Stub value used internally to represent set membership */
const STUB_VALUE: i32 = -1

/**
 * A persistent set collection that stores unique values in the Weil runtime state.
 * 
 * @template T - The type of values in the set
 */
@json
export class WeilSet<T> {
  /** The unique identifier for this set's state storage */
  stateId: WeilId

  /**
   * Creates a new WeilSet instance with the specified state ID.
   * 
   * @param stateId - The unique identifier for this set's state storage
   */
  constructor(stateId: WeilId) {
    this.stateId = stateId
  }

  /**
   * Gets the base state path for this set.
   * 
   * @returns The string representation of the state ID
   */
  private getBaseStatePath(): string {
    return this.stateId.toString()
  }

  /**
   * Generates a unique state tree key for a given value.
   * 
   * @param value - The value to generate a state tree key for
   * @returns A unique string key for state storage
   */
  private getStateTreeKey(value: T): string {
    return [this.getBaseStatePath(), JSON.stringify(value)].join('_')
  }

  /**
   * Adds a value to the set.
   * 
   * @param value - The value to add to the set
   */
  add(value: T): void {
    return Memory.writeCollection<i32>(this.getStateTreeKey(value), STUB_VALUE)
  }

  /**
   * Checks whether a value exists in the set.
   * 
   * @param value - The value to check
   * @returns True if the value exists in the set, false otherwise
   */
  has(value: T): boolean {
    return Memory.readCollection<T>(this.getStateTreeKey(value)) !== null
  }

  /**
   * Removes a value from the set.
   * 
   * @param value - The value to remove from the set
   * @returns The value that was removed, or null if not found
   */
  delete(value: T): T | null {
    return Memory.deleteCollection<T>(this.getStateTreeKey(value))
  }
}
